sync/src/relayer/compact_block_verifier.rs

```diff
@@ -0,0 +1,94 @@
+use crate::relayer::compact_block::CompactBlock;
+use crate::relayer::error::Error;
+use ckb_protocol::{short_transaction_id, short_transaction_id_keys};
+use std::collections::HashSet;
+
+pub struct CompactBlockVerifier {
+    prefilled: PrefilledVerifier,
+    short_ids: ShortIdsVerifier,
+}
+
+impl CompactBlockVerifier {
+    pub(crate) fn new() -> Self {
+        Self {
+            prefilled: PrefilledVerifier::new(),
+            short_ids: ShortIdsVerifier::new(),
+        }
+    }
+
+    pub(crate) fn verify(&self, block: &CompactBlock) -> Result<(), Error> {
+        self.prefilled.verify(block)?;
+        self.short_ids.verify(block)?;
+        Ok(())
+    }
+}
+
+pub struct PrefilledVerifier {}
+
+impl PrefilledVerifier {
+    pub(crate) fn new() -> Self {
+        Self {}
+    }
+
+    pub(crate) fn verify(&self, block: &CompactBlock) -> Result<(), Error> {
+        let prefilled_transactions = &block.prefilled_transactions;
+        let short_ids = &block.short_ids;
+        let txs_len = prefilled_transactions.len() + short_ids.len();
+
+        // Check indices order of prefilled transactions
+        let mut prev = prefilled_transactions
+            .get(0)
+            .and_then(|it| Some(it.index))
+            .unwrap_or(0);
+        for index_transaction in prefilled_transactions.iter().skip(1) {
+            if prev >= index_transaction.index {
+                return Err(Error::UnorderedPrefilledTransactions);
+            }
+            prev = index_transaction.index;
+        }
+
+        // Check highest prefilled index is less then length of block transactions
+        if prev >= txs_len {
+            return Err(Error::OverflowPrefilledTransactions);
+        }
+
+        Ok(())
+    }
+}
+
+pub struct ShortIdsVerifier {}
+
+impl ShortIdsVerifier {
+    pub(crate) fn new() -> Self {
+        Self {}
+    }
+
+    pub(crate) fn verify(&self, block: &CompactBlock) -> Result<(), Error> {
+        let prefilled_transactions = &block.prefilled_transactions;
+        let short_ids = &block.short_ids;
+        let short_ids_set: HashSet<[u8; 6]> = short_ids.iter().map(Clone::clone).collect();
```

Comment:
Can we use `HashSet<&[u8; 6]>` here to remove clone?

Reply:
Changed to `HashSet<&[u8; 6]>`